src/formats/CML.cpp

```diff
@@ -0,0 +1,581 @@
+            }
+        }
+
+        auto new_atom = Atom(title, element);
+        Vector3D cart, velo;
+
+        if (crystal_node && (xf != 0.0 || yf != 0.0 || zf != 0.0)) {
+            cart = frame.cell().matrix() * Vector3D(xf, yf, zf);
+        } else if (x3 == 0.0 && y3 == 0.0 && z3 == 0.0) {
+            cart = Vector3D(x2, y2, 0);
+        } else {
+            cart = Vector3D(x3, y3, z3);
+        }
+
+        // Set special attributes
+        if (formal_charge != 0.0) {
+            new_atom.set_charge(formal_charge);
+        }
+
+        if (isotope != 0.0) {
+            new_atom.set_mass(isotope);
+        }
+
+        if (hydrogen_count != 0.0) {
+            new_atom.set("hydrogen_count", hydrogen_count);
+        }
+
+        for (auto scalar : atom.children("scalar")) {
+            read_property_(new_atom, scalar);
+        }
+
+        // Vectors are stored separate where the values of the vector are
+        // stored as doubles separated by whitespace.
+        for (auto vector3 : atom.children("vector3")) {
+            auto title_attribute = vector3.attribute("title");
+
+            if (!title_attribute) {
+                warning("[CML] vector3 is not titled.");
+                continue;
+            }
+            std::string vec_title = title_attribute.as_string();
+
+            auto vect_strings = split(trim(vector3.text().as_string()), ' ');
+            if (vect_strings.size() != 3) {
+                warning("[CML] {} vector3 does not have 3 values.", vec_title);
+                continue;
+            }
+
+            Vector3D vect;
+            
+            try {
+                vect = { parse<double>(vect_strings[0]),
+                         parse<double>(vect_strings[1]),
+                         parse<double>(vect_strings[2])
+                       };
+            } catch (const Error&) {
+                warning("[CML] {} contains elements which are not numberic.", vec_title);
+                continue;
+            }
+
+            if (vec_title == "velocity") {
+                velo = vect;
+                frame.add_velocities();
+                continue;
+            }
+
+            new_atom.set(vec_title, vect);
+        }
+
+        frame.add_atom(std::move(new_atom), cart, velo);
+        ref_to_id[id] = frame.size() - 1;
+    }
+
+    for (auto bond : current.child("bondArray").children("bond")) {
+        auto atomref = bond.attribute("atomRefs2");
+        auto order = bond.attribute("order");
+
+        if (!atomref) {
+            warning("[CML] Bad bond");
+            continue;
+        }
+
+        auto ids = split(atomref.as_string(), ' ');
+        if (ids.size() != 2) {
+            warning("[CML] Bad bond size");
+            continue;
+        }
+
+        auto id1 = ref_to_id.find(ids[0]);
+        auto id2 = ref_to_id.find(ids[1]);
+        if (id1 == ref_to_id.end() || id2 == ref_to_id.end()) {
+            warning("[CML] Bad bond ref");
```

Comment:
Again, could this be modified to read like `[CML] invalid atomic indexes in bond: {} -- {}`